src/core/Akka.Streams/Dsl/LastElement.cs

```diff
@@ -0,0 +1,73 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="LastElement.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2018 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2013-2018 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System.Threading.Tasks;
+using Akka.Streams.Stage;
+using Akka.Streams.Util;
+
+namespace Akka.Streams.Dsl
+{
+    /// <summary>
+    /// This stage materializes to the last element pushed before upstream completion, if any, thereby recovering from any
+    /// failure. Pushed elements are just passed along.
+    /// </summary>
+    /// <typeparam name="T">input and output type</typeparam>
+    public class LastElement<T> : GraphStageWithMaterializedValue<FlowShape<T, T>, Task<Option<T>>>
+    {
+        #region Logic
+
+        private sealed class Logic : GraphStageLogic
+        {
+            public Logic(LastElement<T> lastElement, TaskCompletionSource<Option<T>> completion) : base(lastElement.Shape)
+            {
+                var currentElement = Option<T>.None;
+
+                SetHandler(lastElement.In, onPush: () =>
+                {
+                    var element = Grab(lastElement.In);
+                    currentElement = new Option<T>(element);
+                    Push(lastElement.Out, element);
+                }, onUpstreamFinish: () =>
+                {
+                    completion.SetResult(currentElement);
+                    CompleteStage();
+                }, onUpstreamFailure: ex =>
+                {
+                    completion.SetResult(currentElement);
```

Comment:
Shouldn’t that be SetException?

Reply:
in Scala version they are not failing stage but completing it upon exception:
https://github.com/akka/akka-stream-contrib/blob/ea7a0ce0dd320a2d3c443fa6097a517b20b31c13/contrib/src/main/scala/akka/stream/contrib/LastElement.scala#L63

this is also captured in the unit test ```materialize to the last element emitted by a source before it failed```:
https://github.com/akka/akka-stream-contrib/blob/ab0817a660906316e282d0c7c762cb1d18a05c5d/contrib/src/test/scala/akka/stream/contrib/LastElementSpec.scala#L41

Reply:
@alex-bogomaz alright, thx for the clarification, will merge once build is green 👍